drivers/clock_control/nrf_power_clock.c

```diff
@@ -226,68 +188,157 @@ static inline void anomaly_132_workaround(void)
 #endif
 }
 
-static int clock_async_start(struct device *dev,
-			     clock_control_subsys_t subsys,
-			     struct clock_control_async_data *data)
+static int async_start(struct device *dev, clock_control_subsys_t subsys,
+			struct clock_control_async_data *data, uint32_t ctx)
 {
 	enum clock_control_nrf_type type = (enum clock_control_nrf_type)subsys;
-	const struct nrf_clock_control_sub_config *config;
-	struct nrf_clock_control_sub_data *clk_data;
-	int key;
-	uint8_t ref;
+	struct nrf_clock_control_sub_data *subdata = get_sub_data(dev, type);
+	int err;
 
-	__ASSERT_NO_MSG(type < CLOCK_CONTROL_NRF_TYPE_COUNT);
-	config = get_sub_config(dev, type);
-	clk_data = get_sub_data(dev, type);
-
-	__ASSERT_NO_MSG((data == NULL) ||
-			((data != NULL) && (data->cb != NULL)));
-
-	/* if node is in the list it means that it is scheduled for
-	 * the second time.
-	 */
-	if ((data != NULL)
-	    && is_in_list(&clk_data->list, &data->node)) {
-		return -EBUSY;
+	err = set_state(&subdata->flags, CLOCK_CONTROL_STATUS_OFF,
+			ctx | CLOCK_CONTROL_STATUS_STARTING);
+	if (err < 0) {
+		return err;
 	}
 
-	key = irq_lock();
-	ref = ++clk_data->ref;
-	__ASSERT_NO_MSG(clk_data->ref > 0);
-	irq_unlock(key);
+	subdata->cb = data->cb;
+	subdata->user_data = data->user_data;
 
-	if (data) {
-		bool already_started;
+	if (IS_ENABLED(CONFIG_NRF52_ANOMALY_132_WORKAROUND) &&
+		(subsys == CLOCK_CONTROL_NRF_SUBSYS_LF)) {
+		anomaly_132_workaround();
+	}
 
-		clock_irqs_disable();
-		already_started = clk_data->started;
-		if (!already_started) {
-			list_append(&clk_data->list, &data->node);
-		}
-		clock_irqs_enable();
+	nrf_clock_task_trigger(NRF_CLOCK, get_sub_config(dev, type)->start_tsk);
 
-		if (already_started) {
-			data->cb(dev, subsys, data->user_data);
-		}
+	return 0;
+}
+
+static int api_start(struct device *dev, clock_control_subsys_t subsys,
+			     struct clock_control_async_data *data)
+{
+	return async_start(dev, subsys, data, CTX_API);
+}
+
+static void blocking_start_callback(struct device *dev,
+				    clock_control_subsys_t subsys,
+				    void *user_data)
+{
+	struct k_sem *sem = user_data;
+
+	k_sem_give(sem);
+}
+
+static int api_blocking_start(struct device *dev, clock_control_subsys_t subsys)
+{
+	struct k_sem sem = Z_SEM_INITIALIZER(sem, 0, 1);
+	struct clock_control_async_data data = {
+		.cb = blocking_start_callback,
+		.user_data = &sem
+	};
+	int err;
+
+	err = api_start(dev, subsys, &data);
+	if (err < 0) {
+		return err;
 	}
 
-	if (ref == 1) {
-		DBG(dev, subsys, "Triggering start task");
+	return k_sem_take(&sem, K_MSEC(500));
```

Comment:
`clock_control_on` in 'clock_control.h' states
```
This function may sleep, and thus can only be called from thread context.
```
For me it means that it can only be called from app which supports multithreading.

Reply:
We could extend description to state that in non sleepable context it just spins and then extend implementation.

Reply:
I think that "from thread context" in that description is supposed to mean "not from ISR". And someone may want to call this function from `main()` in an app without `MULTITHREADING` enabled. An attempt to call `k_sem_take` with a parameter other than `K_NO_WAIT` will end up in UsageFault then (I thought it would just not wait but apparently I was wrong). I think this case should be handled somehow, perhaps with a loop calling `k_cpu_idle()` while waiting for this semaphore.

Reply:
just learn that interrupts does not work if multhreading is off (#26796). I will return error in that case.